### DISCUSSION TITLE
Why is Semver option not working in flux2

### DISCUSSION THREAD

**Discussion by Ashik01** [fluxcd/flux2#4566] [General]
I am using flux2 version 2.1.2 , Under git repository semver option is not working . So I have move to branch option instead . 

When I try to semver in below repository.yaml instead of branch it is not working as expected

spec:
  interval: 10m
  ref:
    **branch: xyz**
  secretRef:
    name: git-credentials
  timeout: 1m0s

Error 1 :  Tried ALPHA Numerical tag

NAMESPACE       NAME                            REVISION        SUSPENDED       READY   MESSAGE                                                              
kube-system     gitrepository/eks-addons                        False           False   failed to checkout and determine revision: semver parse error: improper constraint: ~AD01.01.08

Error 2 : Numerical Tag 

NAMESPACE       NAME                            REVISION        SUSPENDED       READY   MESSAGE                                                              
kube-system     gitrepository/eks-addons                        False           False   failed to checkout and determine revision: unable to resolve tag revision: unsupported object type

**Comment by kingdonb** [comment]
Please post the YAML that didn't work, we can only diagnose it if we can see what you tried

Also, please use triple backticks to format your code blocks:

```
   \```
```

That way we can see better the content, it's important to preserve the indentation, otherwise yaml is basically unusable

**Reply by kingdonb** [reply]
I see at least one error trying to parse through your posting @Ashik01 - semver tags cannot have leading zeroes. If your semver expression looks like 01.01.08 then it is not valid semver. You can still use these tags with alphabetic sort, maybe, but I'd be very careful with that. It would be better to implement semver properly in your tagging system, then you will likely see the issue abated. (See https://semver.org for the full semver spec which Flux implements faithfully)

**Comment by Ashik01** [comment]
Thanks for response @kingdonb , Please find the YML below 

```
apiVersion: source.toolkit.fluxcd.io/v1beta2
kind: GitRepository
metadata:
  annotations:
  finalizers:
  - finalizers.fluxcd.io
  generation: 2
  name: eks-addons
  namespace: kube-system
  resourceVersion: "70621"
  uid: cb8d9208-a2bd-
spec:
  gitImplementation: go-git
  interval: 10m0s
  ref:
    semver: '01.01.08'
  secretRef:
    name: git-credentials
  timeout: 1m0s
  url: ssh://git@gitlab.verizon.com/******
status:
  artifact:
    checksum: 4bd853ca11159c74b629b9b446d4e
    lastUpdateTime: "2024-01-29T08:01:21Z"
    path: gitrepository/*********
    revision: release/***************
    url: ********************
  conditions:
  - lastTransitionTime: "2024-01-29T10:01:22Z"
    message: stored artifact for revision ''
    observedGeneration: 2
    reason: Succeeded
    status: "True"
    type: Ready
  - lastTransitionTime: "2024-01-29T08:01:21Z"
    message: stored artifact for revision ''
    observedGeneration: 2
    reason: Succeeded
    status: "True"
    type: ArtifactInStorage
  lastHandledReconcileAt: "2024-01-29T10:15:07.975402641Z"
  observedGeneration: 2
  url: 
```

**Reply by kingdonb** [reply]
As I suspected,

> semver tags cannot have leading zeroes. If your semver expression looks like 01.01.08 then it is not valid semver.

It looks like your semver expression is not valid semver. We could perhaps do better on the error message, but it seems relatively clear this is pointed at from the original error that you posted:

> improper constraint: ~AD01.01.08

As long as you're not trying to use a wildcard here, it would make sense to use the `ref.tag` instead of `ref.semver` - it has no constraints on whether the tags must be valid semver or not. But again, I would suggest to adjust your tags to follow semver, since it seems the intention of providing a dotted three-ple was to implement semver. If you use a valid semver, then you can take advantage of Flux's wildcard support for ranges and otherwise!

Hope this helps
